sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/continuous/ContinuousRateStreamSource.scala

```diff
@@ -82,7 +87,16 @@ class RateStreamContinuousStream(rowsPerSecond: Long, numPartitions: Int) extend
     RateStreamContinuousReaderFactory
   }
 
-  override def commit(end: Offset): Unit = {}
+  override def commit(end: Offset): Unit = {
+    end.asInstanceOf[RateStreamOffset].partitionToValueAndRunTimeMs.foreach {
+      case (_, ValueRunTimeMsPair(value, _)) =>
+        if (highestCommittedValue.get() < value) {
```

Comment:
I'm not sure if the 'atomic' part is essential here, but if it is, I think you have a race condition here. You'd want to use `updateAndGet` or something to make sure the check and update are atomic.

Reply:
Only one writer and one reader will run concurrently. I'm revisiting the change, and it looks like just over-engineering. `volatile` would just work. Thanks for pointing out!